components/zcash_protocol/src/consensus.rs

```diff
@@ -677,6 +701,39 @@ impl BranchId {
         !matches!(self, BranchId::Sprout | BranchId::Overwinter)
     }
 }
+#[allow(unused_variables)]
+fn upper_bound<P: Parameters>(target_height: BlockHeight, params: &P) -> Option<BlockHeight> {
+    #[allow(dead_code)]
+    fn check_bound<P: Parameters>(
+        target_height: BlockHeight,
+        current_upper: Option<BlockHeight>,
+        nu: NetworkUpgrade,
+        params: &P,
+    ) -> Option<BlockHeight> {
+        let nu_height = params.activation_height(nu);
+        if nu_height.is_some() && nu_height.unwrap() > target_height {
+            nu_height
+        } else {
+            current_upper
+        }
+    }
+
+    #[allow(unused_mut)]
+    let mut upper = None;
+    #[cfg(zcash_unstable = "nu6")]
+    {
+        upper = check_bound(target_height, upper, NetworkUpgrade::Nu6, params);
+    }
+    #[cfg(zcash_unstable = "nu6" /* TODO nu7 */ )]
+    {
+        upper = check_bound(target_height, upper, NetworkUpgrade::Nu7, params);
+    }
+    #[cfg(zcash_unstable = "zfuture")]
+    {
+        upper = check_bound(target_height, upper, NetworkUpgrade::ZFuture, params);
+    }
+    upper
+}
```

Comment:
It will not work like this. For example, if we build with nu7 flag, but target_height is between nu5 and nu6 - this will return nu7 rather than nu6

But with your suggestion about method simplification it already got better

If we we allow experimental feature and remove brackets - it will be very smooth